conveyor/slot.go

```diff
@@ -0,0 +1,238 @@
+	pulseState            PulseState
+	slotState             SlotState
+	pulse                 core.Pulse
+	pulseNumber           core.PulseNumber
+	nodeID                uint32
+	nodeData              interface{}
+	elements              []slotElement
+	elementListMap        map[ActivationStatus]*ElementList
+}
+
+// SlotStateMachine represents state machine of slot itself
+var SlotStateMachine = slotElement{
+	id:               0,
+	state:            0,
+	stateMachineType: 0,
+}
+
+func initElementsBuf() []slotElement {
+	elements := make([]slotElement, slotSize)
+	elements[0] = SlotStateMachine
+	var nextElement *slotElement
+	for i := slotSize - 1; i > 0; i-- {
+		elements[i] = *newSlotElement(EmptyElement)
+		elements[i].id = uint32(i)
+		elements[i].listNext = nextElement
+		nextElement = &elements[i]
+	}
+	return elements
+}
+
+// NewSlot creates new instance of Slot
+func NewSlot(pulseState PulseState, pulseNumber core.PulseNumber) *Slot {
+	slotState := Initializing
+	if pulseState == Antique {
+		slotState = Working
+	}
+
+	elements := initElementsBuf()
+
+	empty := &ElementList{
+		// elements[0] contains SlotStateMachine, so first empty element is elements[1]
+		head: &elements[1],
+		tail: &elements[slotSize-1],
+	}
+
+	elementListMap := map[ActivationStatus]*ElementList{
+		EmptyElement:     empty,
+		ActiveElement:    &ElementList{},
+		NotActiveElement: &ElementList{},
+	}
+	return &Slot{
+		pulseState:     pulseState,
+		inputQueue:     queue.NewMutexQueue(),
+		pulseNumber:    pulseNumber,
+		slotState:      slotState,
+		elements:       elements,
+		elementListMap: elementListMap,
+	}
+}
+
+func (s *Slot) getPulseNumber() core.PulseNumber {
+	return s.pulseNumber
+}
+
+func (s *Slot) getPulseData() core.Pulse {
+	return s.pulse
+}
+
+func (s *Slot) getNodeID() uint32 {
+	return s.nodeID
+}
+
+func (s *Slot) getNodeData() interface{} {
+	return s.nodeData
+}
+
+// createElement creates new active element from empty element
+func (s *Slot) createElement(stateMachineType StateMachineType, state uint16, event queue.OutputElement) *slotElement {
+	element := s.popElement(EmptyElement)
+	element.stateMachineType = stateMachineType
+	element.state = state
+	element.activationStatus = ActiveElement
+	element.listNext = nil
+	// Set other fields to element, like:
+	// element.payload = event.GetPayload()
+
+	s.pushElement(ActiveElement, element)
+	return element
+}
+
+// popElement gets element of provided status from correspondent linked list (and remove it from that list)
+func (s *Slot) popElement(status ActivationStatus) *slotElement {
+	list := s.elementListMap[status]
+	return list.popElement()
+}
+
+// pushElement adds element of provided status to correspondent linked list
+func (s *Slot) pushElement(status ActivationStatus, element *slotElement) {
+	element.activationStatus = status
+	if status == EmptyElement {
+		element.id = element.id + slotElementDelta
+	}
+	list := s.elementListMap[status]
```

Comment:
what if there is no 'status' in s.elementListMap?

Reply:
We create list for every ActivationStatus when create new slot